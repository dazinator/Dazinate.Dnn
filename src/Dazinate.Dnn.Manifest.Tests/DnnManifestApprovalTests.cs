using System;
using System.IO;
using System.Linq;
using System.Text;
using System.Xml;
using ApprovalTests;
using ApprovalTests.Reporters;
using Autofac;
using Dazinate.Dnn.Manifest.Factory;
using Dazinate.Dnn.Manifest.Ioc;
using Dazinate.Dnn.Manifest.Package.Dependency;
using Xunit;
using Dazinate.Dnn.Manifest.Package.Component.Assembly;
using Dazinate.Dnn.Manifest.Package.Component.AuthenticationSystem;
using Dazinate.Dnn.Manifest.Package.Component.Cleanup;
using Dazinate.Dnn.Manifest.Package.Component.Shared.File;

namespace Dazinate.Dnn.Manifest.Tests
{
    [UseReporter(typeof(DiffReporter))]
    [Collection("Csla")]
    public class DnnManifestApprovalTests : BaseBusinessTest, IDisposable
    {

        /// <summary>
        /// Constructor is executed prior to every individual test.
        /// </summary>
        public DnnManifestApprovalTests()
        {
            Console.Write("initialising");
        }

        private string LoadManifestXml(string localFileName)
        {
            var dir = System.IO.Directory.GetCurrentDirectory();
            var filePath = Path.Combine(dir, localFileName);
            var xmlContents = System.IO.File.ReadAllText(filePath);
            return xmlContents;
        }


        [Theory]
        [InlineData("manifest.xml")]
        public void Can_Load_And_Save_Manifest(string manifestFile)
        {
            //  var names = new[] { "Llewellyn", "James", "Dan", "Jason", "Katrina" };
            // Array.Sort(names);
            var xmlContents = LoadManifestXml(manifestFile);
            IDnnManifestFactory<IPackagesDnnManifest> factory = new PackagesDnnManifestFactory();

            // Act           
            var dnnManifest = factory.Get(xmlContents);
            dnnManifest.Packages[0].Description = "changed";

            var xmlStringBuilder = new StringBuilder();
            using (XmlWriter xmlWriter = XmlWriter.Create(new StringWriter(xmlStringBuilder)))
            {
                dnnManifest = (IPackagesDnnManifest)dnnManifest.SaveToXml(xmlWriter);
            }

            // Now verify the xml looks good.
            Approvals.VerifyXml(xmlStringBuilder.ToString());
        }

        [Fact]
        public void Can_Create_Manifest()
        {

            var factory = new PackagesDnnManifestFactory();

            // Act   
            // Create a fully populated package manifest, demonstrating all possible manifest features.        
            var dnnManifest = factory.CreateNewManifest();
            dnnManifest.Version = "5.0";
            dnnManifest.Type = ManifestType.Package;

            // Add an Auth_System package.
            var authSystemPackage = dnnManifest.Packages.AddNewPackage();
            authSystemPackage.Name = "MyAuthSystemPackage";
            authSystemPackage.Description = "An amazing auth system.";
            authSystemPackage.FriendlyName = "My Amazing Auth System";
            authSystemPackage.Version = "1.0.0";
            authSystemPackage.Type = "Auth_System";

            // The auth system package has some dependencies.
            var dependency = (CoreVersionDependency)authSystemPackage.Dependencies.AddNewCoreVersionDependency();
            dependency.Version = "5.0";

            var customDependency = (CustomDependency)authSystemPackage.Dependencies.AddNewCustomDependency();
            customDependency.Type = "Custom";
            customDependency.Value = "SpecialValue";

            var managedPackagedDependency = (ManagedPackageDependency)authSystemPackage.Dependencies.AddNewManagedPackageDependency();
            managedPackagedDependency.PackageName = "SomeOtherPackage";
            managedPackagedDependency.Version = "1.0.0";

            var packagedDependency = (PackageDependency)authSystemPackage.Dependencies.AddNewPackageDependency();
            packagedDependency.PackageName = "AndAnotherPackage";

            var typeDependency = (TypeDependency)authSystemPackage.Dependencies.AddNewTypeDependency();
            typeDependency.TypeName = "System.Version";

            var xmlStringBuilder = new StringBuilder();
            using (XmlWriter xmlWriter = XmlWriter.Create(new StringWriter(xmlStringBuilder)))
            {
                dnnManifest = (IPackagesDnnManifest)dnnManifest.SaveToXml(xmlWriter);
            }

            // Add an assembly component with some assemblies listed.
            var assyComponent = authSystemPackage.Components.AddNewComponent<IAssemblyComponent>();
            IAssembly assy = (IAssembly)assyComponent.Assemblies.AddNew();
            assy.Name = "test.dll";
            assy.Path = "bin";
            //  assyComponent.Assemblies.ad

            // Now verify the xml looks good.
            Approvals.VerifyXml(xmlStringBuilder.ToString());
        }

        [Fact]
        public void Can_Add_AssemblyComponent()
        {

            var factory = new PackagesDnnManifestFactory();

            // Act   
            // Create a fully populated package manifest, demonstrating all possible manifest features.        
            var dnnManifest = factory.CreateNewManifest();
            dnnManifest.Version = "5.0";
            dnnManifest.Type = ManifestType.Package;

            // Add an Auth_System package.
            var authSystemPackage = dnnManifest.Packages.AddNewPackage();
            authSystemPackage.Name = "MyAuthSystemPackage";
            authSystemPackage.Description = "An amazing auth system.";
            authSystemPackage.FriendlyName = "My Amazing Auth System";
            authSystemPackage.Version = "1.0.0";
            authSystemPackage.Type = "Auth_System";

            // Add an assembly component with some assemblies listed.
            var assyComponent = authSystemPackage.Components.AddNewComponent<IAssemblyComponent>();
            IAssembly assy = (IAssembly)assyComponent.Assemblies.AddNew();
            assy.Name = "foo.dll";
            assy.Path = "bin";

            IAssembly anotherAssy = (IAssembly)assyComponent.Assemblies.AddNew();
            anotherAssy.Name = "bar.dll";
            anotherAssy.Path = "bin";
            anotherAssy.Version = "1.0.0";
            anotherAssy.Action = AssemblyAction.Install;

            IAssembly anotherAssyUnregister = (IAssembly)assyComponent.Assemblies.AddNew();
            anotherAssyUnregister.Name = "baz.dll";
            anotherAssyUnregister.Path = "bin";
            anotherAssyUnregister.Version = "1.0.0.1";
            anotherAssyUnregister.Action = AssemblyAction.Unregister;


            var xmlStringBuilder = new StringBuilder();
            using (XmlWriter xmlWriter = XmlWriter.Create(new StringWriter(xmlStringBuilder)))
            {
                dnnManifest = (IPackagesDnnManifest)dnnManifest.SaveToXml(xmlWriter);
            }
            // Now verify the xml looks good.
            Approvals.VerifyXml(xmlStringBuilder.ToString());
        }

        [Fact]
        public void Can_Add_AuthenticationSystemComponent()
        {

            var factory = new PackagesDnnManifestFactory();

            // Act   
            // Create a fully populated package manifest, demonstrating all possible manifest features.        
            var dnnManifest = factory.CreateNewManifest();
            dnnManifest.Version = "5.0";
            dnnManifest.Type = ManifestType.Package;

            // Add an Auth_System package.
            var authSystemPackage = dnnManifest.Packages.AddNewPackage();
            authSystemPackage.Name = "MyAuthSystemPackage";
            authSystemPackage.Description = "An amazing auth system.";
            authSystemPackage.FriendlyName = "My Amazing Auth System";
            authSystemPackage.Version = "1.0.0";
            authSystemPackage.Type = "Auth_System";

            // Add an assembly component with some assemblies listed.
            var authComponent = authSystemPackage.Components.AddNewComponent<IAuthenticationSystemComponent>();
            authComponent.Type = "foo";
            authComponent.LoginControlSource = "/some/login.ascx";
            authComponent.LogoffControlSource = "/some/logoff.ascx";
            authComponent.SettingsControlSource = "/some/settings.ascx";

            var xmlStringBuilder = new StringBuilder();
            using (XmlWriter xmlWriter = XmlWriter.Create(new StringWriter(xmlStringBuilder)))
            {
                dnnManifest = (IPackagesDnnManifest)dnnManifest.SaveToXml(xmlWriter);
            }
            // Now verify the xml looks good.
            Approvals.VerifyXml(xmlStringBuilder.ToString());
        }

        [Fact]
        public void Can_Add_CleanupComponent()
        {

            var factory = new PackagesDnnManifestFactory();

            // Act   
            // Create a fully populated package manifest, demonstrating all possible manifest features.        
            var dnnManifest = factory.CreateNewManifest();
            dnnManifest.Version = "5.0";
            dnnManifest.Type = ManifestType.Package;

            // Add an Auth_System package.
            var authSystemPackage = dnnManifest.Packages.AddNewPackage();
            authSystemPackage.Name = "MyAuthSystemPackage";
            authSystemPackage.Description = "An amazing auth system.";
            authSystemPackage.FriendlyName = "My Amazing Auth System";
            authSystemPackage.Version = "1.0.0";
            authSystemPackage.Type = "Auth_System";

            // Add an assembly component with some assemblies listed.
            var component = authSystemPackage.Components.AddNewComponent<ICleanupComponent>();

            var file = (IFile)component.Files.AddNew();
            file.Name = "foo.txt";
            file.Path = "files";
            file.SourceFileName = "bar.txt";

            var xmlStringBuilder = new StringBuilder();
            using (XmlWriter xmlWriter = XmlWriter.Create(new StringWriter(xmlStringBuilder)))
            {
                dnnManifest = (IPackagesDnnManifest)dnnManifest.SaveToXml(xmlWriter);
            }
            // Now verify the xml looks good.
            Approvals.VerifyXml(xmlStringBuilder.ToString());
        }



    }


}


