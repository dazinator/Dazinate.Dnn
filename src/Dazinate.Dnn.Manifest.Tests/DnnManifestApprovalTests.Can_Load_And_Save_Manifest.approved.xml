<dotnetnuke type="Package" version="6.0">
  <packages>
    <package name="SomePackageName" type="Module" version="0.0.1">
      <friendlyName>SomePackageName</friendlyName>
      <description>changed</description>
      <owner>
        <name>Daz</name>
        <organization>Daz</organization>
        <url>http://www.someurl.com</url>
        <email>&lt;a href="mailto:support@someorg.com"&gt;support@someorg.com&lt;/a&gt;</email>
      </owner>
      <license src="License.lic" />
      <releaseNotes src="ReleaseNotes.txt" />
      <dependencies>
        <dependency type="coreVersion">05.04.01</dependency>
        <dependency type="managedPackage" version="1.0.0">AnotherPackageRequiredByThisComponent</dependency>
        <dependency type="package">AnotherPackageRequiredByThisComponent</dependency>
        <dependency type="type">System.Security.Principal.GenericPrincipal</dependency>
        <dependency type="SomeCustomDependencyType">ValueNeededBySomeCustomDependencyType</dependency>
      </dependencies>
      <components>
        <component type="Assembly">
          <assemblies>
            <assembly>
              <path>bin</path>
              <name>[YourAssembly.dll]</name>
              <version>1.0.0.0</version>
            </assembly>
            <assembly action="Unregister">
              <path>bin</path>
              <name>SomeOld.dll</name>
              <version>1.0.0.0</version>
            </assembly>
          </assemblies>
        </component>
        <component type="AuthenticationSystem">
          <authenticationService>
            <type>OpenID</type>
            <settingsControlSrc>DesktopModules/AuthenticationServices/OpenID/Settings.ascx</settingsControlSrc>
            <loginControlSrc>DesktopModules/AuthenticationServices/OpenID/Login.ascx</loginControlSrc>
            <logoffControlSrc />
          </authenticationService>
        </component>
        <component type="Cleanup" version="07.40.00">
          <files>
            <file>
              <path>bin</path>
              <name>MyFile.dll</name>
            </file>
            <file>
              <path>bin</path>
              <name>MyOtherFile.dll</name>
            </file>
          </files>
        </component>
        <component type="Config">
          <config>
            <configFile>web.config</configFile>
            <install>
              <configuration>
                <nodes>
                  <node path="/configuration/system.webServer/handlers" action="update" key="path" collision="overwrite">
                    <add name="RadProgressHandler" verb="*" path="Telerik.RadUploadProgressHandler.ashx" type="Telerik.Web.UI.Upload.RadUploadProgressHandler, Telerik.Web.UI" preCondition="integratedMode" />
                  </node>
                  <node />
                </nodes>
              </configuration>
            </install>
            <uninstall>
              <configuration>
                <nodes />
              </configuration>
            </uninstall>
          </config>
        </component>
        <component type="Container">
          <containerFiles>
            <basePath />
            <containerName>MyContainer</containerName>
            <containerFile>
              <path>somefile</path>
              <name>somename</name>
            </containerFile>
            <containerFile>
              <path />
              <name />
            </containerFile>
          </containerFiles>
        </component>
        <component type="DashboardControl">
          <dashboardControl>
            <key>Server</key>
            <src>DesktopModules/Admin/Dashboard/Modules/Server.ascx</src>
            <localResources>DesktopModules/Admin/Dashboard/App_LocalResources/SharedResources.ascx</localResources>
            <controllerClass>DotNetNuke.Modules.Dashboard.Components.Server.ServerController</controllerClass>
            <isEnabled>true</isEnabled>
            <viewOrder>1</viewOrder>
          </dashboardControl>
        </component>
        <component type="File">
          <files>
            <basePath>DesktopModules\MyModule</basePath>
            <file>
              <path>app_localresources</path>
              <name>MyModule.ascx.resx</name>
            </file>
            <file>
              <path />
              <name>MyModule.ascx</name>
              <sourceFileName>some/path/in/zip/mymodule.ascx</sourceFileName>
            </file>
          </files>
        </component>
        <component type="CoreLanguage">
          <languageFiles>
            <code>ca-ES</code>
            <displayName>Catala (España)</displayName>
            <fallback>en-US</fallback>
            <languageFile>
              <path>admin\ControlPanel\App_LocalResources</path>
              <name>classic.ascx.ca-es.resx</name>
            </languageFile>
            <languageFile>
              <path>admin\ControlPanel\App_LocalResources</path>
              <name>iconbar.ascx.ca-es.resx</name>
            </languageFile>
          </languageFiles>
        </component>
        <component type="ExtensionLanguage">
          <languageFiles>
            <code>en-US</code>
            <package>DefaultAuthentication</package>
            <basePath>DesktopModules\AuthenticationServices\DNN</basePath>
            <languageFile>
              <path>app_localresources</path>
              <name>Login.ascx.resx</name>
            </languageFile>
            <languageFile>
              <path>app_localresources</path>
              <name>Settings.ascx.resx</name>
            </languageFile>
          </languageFiles>
        </component>
      </components>
    </package>
  </packages>
</dotnetnuke>